cuqi/experimental/mcmc/_sampler.py

```diff
@@ -104,6 +117,25 @@ def target(self, value):
         self._target = value
         self.validate_target()
 
+    @property
+    def current_point(self):
+        """ The current point of the sampler. """
+        return self._current_point
+    
+    @current_point.setter
+    def current_point(self, value):
+        """ Set the current point of the sampler. """
+        self._current_point = value
+
+    @property
+    def current_target_logd(self):
+        """ The current target value of the sampler. """
+        return self._current_target_logd
+    
+    @current_target_logd.setter
+    def current_target_logd(self, value):
+        """ Set the current target value of the sampler. """
+        self._current_target_logd = value
 
```

Comment:
This comment probably not necessarily should be addressed in this PR. I think we need a way to make sure developers who create subclass of NewSampler do use these two properties. Here are two suggestion:
1- to enforce these as abstractmethods. To prevent a case when developer make a subclass of NewSampler and not consider using these properties at all. That is, it is part of the interface the developer needs to implement. This is I think is a tradeoff. It will lead to some redundancy, but it will make it clear that the developer need to take care of 'step' 'tune' and the two states: 'current_point', ' current_target_logd'.
2- as we did before, ask user to implement `_tune` and `_step` instead of `tune` and `step`. Then we have parent method `tune` and `step` to call `_tune` / `_step` expecting `current_point` and `target_logd` to be returned. and store them in state variable. So it is not the responsibility of  `_tune` / `_step` to store this state variable.

Reply:
I agree there is something to consider here. I went with the following option:

Added a unit test to check if subclassing samplers utilize all state variables. See test_state_is_fully_updated_after_warmup_step.

We could still create an issue if you want to consider future, e.g. if users subclass within their ipynb or something.

We could add a "validate_sampler" method that can be run to check it works as expected.

Reply:
really nice added tests!

I just created this issue #392 for a general user-defined sampler (I do not think we have one). If you think it makes sense, you can add the details you suggested in this issue.